src/pyuvdata/uvbeam/analytic_beam.py

```diff
@@ -0,0 +1,893 @@
+# Copyright (c) 2022 Radio Astronomy Software Group
+# Licensed under the 2-clause BSD License
+
+"""Analytic beam class definitions."""
+
+from __future__ import annotations
+
+import dataclasses
+import importlib
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Literal
+
+import numpy as np
+import numpy.typing as npt
+import yaml
+from astropy.constants import c as speed_of_light
+from scipy.special import j1
+
+from .. import utils
+from ..docstrings import combine_docstrings
+from .uvbeam import UVBeam, _convert_feeds_to_pols
+
+analytic_beam_classes = ["AiryBeam", "GaussianBeam", "ShortDipoleBeam", "UniformBeam"]
+__all__ = ["AnalyticBeam"] + analytic_beam_classes
+
+
+@dataclass
+class AnalyticBeam(ABC):
+    """
+    Define an analytic beam base class.
+
+    Attributes
+    ----------
+    basis_vec_dict : dict
+        Allowed basis vector types (keys) with their number of axes (which will be set
+        on the object as Naxes_vec).
+    basis_vector_type : str
+        The type of basis vectors this beam is defined with. Must be a key in
+        `basis_vec_dict`. Currently, only "az_za" is supported, which means basis
+        vectors aligned with the zenith angle and azimuth.
+    Naxes_vec : int
+        Number of basis vectors. This is set based on the `basis_vector_type`.
+    Nfeeds : int
+        Number of feeds. This is set based on the `feed_array`.
+
+    Parameters
+    ----------
+    feed_array : np.ndarray of str
+        Feeds to define this beam for, e.g. x & y or n & e (for "north" and "east")
+        or r & l.
+    include_cross_pols : bool
+        Option to include the cross polarized beams (e.g. xy and yx or en and ne) for
+        the power beam.
+    x_orientation : str
+        Physical orientation of the feed for the x feed. Not meaningful for
+        unpolarized analytic beams, but clarifies the orientation of the dipole
+        for the ShortDipoleBeam and matches with the meaning on UVBeam objects.
+
+    """
+
+    # In the future, this might allow for cartesian basis vectors in some orientation.
+    # In that case, the Naxes_vec would be 3 rather than 2
+    _basis_vec_dict = {"az_za": 2}
+
+    @property
+    @abstractmethod
+    def basis_vector_type(self):
+        """Require that a basis_vector_type is defined in concrete classes."""
+
+    def __init__(
+        self,
+        *,
+        feed_array: npt.NDArray[str] | None = None,
+        include_cross_pols: bool = True,
+        x_orientation: Literal["east", "north"] = "east",
+    ):
+        if self.basis_vector_type not in self._basis_vec_dict:
+            raise ValueError(
+                f"basis_vector_type is {self.basis_vector_type}, must be one of "
+                f"{list(self._basis_vec_dict.keys())}"
+            )
```

Comment:
since `basis_vector_type` is a _class_ variable, not an instance variable, it should be validated at class construction time (i.e. in the `__init_subclass__` method).

Reply:
oh, sorry I'm wrong -- it's a property. But I think you mean it to be a class attribute?

Reply:
It's a class attribute, but it's required to exist, so you had me write it as a property so that we could use the `abstractmethod` decorator. In principal, it can be different for different subclasses.

Reply:
Seems to work to move it into the `__init_subclass__` and I agree it's cleaner there.